### PR TITLE
raftstore-v2: fix incorrect GC peer requests to source peer after merge

### LINKED ISSUE #15623: [Dynamic Regions] resolved ts can't  advance for long time
### What version of TiKV are you using?

TiKV
Release Version:   7.4.0-alpha
Edition:           Community
Git Commit Hash:   e43a157c4a35034dfd705bdd94fac6d958e8a1ff
Git Commit Branch: heads/refs/tags/v7.4.0-alpha
UTC Build Time:    2023-09-14 11:47:25
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
### Steps to reproduce

The cluster contains about 60 tikv with dynamic region enabled, each tikv contains 2.5TB data, a table contains about 6billion rows with each row size about 10kb.
b8ae02ef-9ddc-4571-aa78-753117c01a19

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #15623

Depends on https://github.com/pingcap/kvproto/pull/1190 

What's Changed:

```commit-message
After merge, target region sends GC peer requests to removed source peers,
however the region_id in requests is set to target region id incorrectly.
As results, source region removed peers may be left forever.

This commit fixes above issue by putting source removed_records to
merged_records, so that region id can be set correctly.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -201,8 +201,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "7693954bd1dd86e
 # When you modify TiKV cooperatively with kvproto, this will be useful to submit the PR to TiKV and the PR to
 # kvproto at the same time.
 # After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
-# [patch.'https://github.com/pingcap/kvproto']
-# kvproto = { git = "https://github.com/your_github_id/kvproto", branch = "your_branch" }
+[patch.'https://github.com/pingcap/kvproto']
+kvproto = { git = "https://github.com/overvenus/kvproto", branch = "fix-merge-gc-peer" }
```
Comment (tonyxuqqi) [inline Cargo.toml:205]: make sure it's updated before merging.

File: components/raftstore-v2/src/operation/command/admin/conf_change.rs [hunk 2]
```diff
@@ -609,10 +609,13 @@ impl<EK: KvEngine, R> Apply<EK, R> {
...
         merged_records.retain_mut(|r| {
-            let mut sources: Vec<_> = r.take_source_peers().into();
-            sources.retain(|p| !updates.contains(&p.get_id()));
-            r.set_source_peers(sources.into());
-            !r.get_source_peers().is_empty()
+            let mut source_peers: Vec<_> = r.take_source_peers().into();
+            source_peers.retain(|p| !updates.contains(&p.get_id()));
+            r.set_source_peers(source_peers.into());
+            let mut source_removed_records: Vec<_> = r.take_source_removed_records().into();
+            source_removed_records.retain(|p| !updates.contains(&p.get_id()));
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/command/admin/conf_change.rs:616]: Could you add some comments on the change

**Review by SpadeA-Tang — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**